### DISCUSSION TITLE
JOIN between different databases AND using different files in a module seems to get SQA confused

### DISCUSSION THREAD

**Discussion by MarkBFamFour** [sqlalchemy/sqlalchemy#8027] [Usage Questions] [answered]
## Stack ##

I am using:

 - Python 3.10.x 
 - FastAPI 0.75.x 
 - SQLAlchemy 1.4.3x

## Summary ##

I am building a unifying FastAPI project for several legacy databases (stored back-end on MariaDB 10.3 - structure has to be retained for some legacy software).

My SQLA setup uses a databases module to do the following:

### /databases.py ###

``` lang-python
import dotenv
import os

from sqlalchemy import create_engine
from sqlalchemy.orm import sessionmaker

import .models as models

dotenv.load_dotenv()

engines = {
    'parts': create_engine("mysql+pymysql://" + os.environ['DB_URL'] + "/parts", pool_pre_ping=True, pool_recycle=300),
    'shop': create_engine("mysql+pymysql://" + os.environ['DB_URL'] + "/shop", pool_pre_ping=True, pool_recycle=300),
    'purchasing': create_engine("mysql+pymysql://" + os.environ['DB_URL'] + "/purchasing", pool_pre_ping=True, pool_recycle=300),
    "company": create_engine("mysql+pymysql://" + os.environ['DB_URL'] + "/company", pool_pre_ping=True, pool_recycle=300),
    "auth": create_engine("mysql+pymysql://" + os.environ['DB_URL'] + "/auth", pool_pre_ping=True, pool_recycle=300),
}

DBSession = sessionmaker(autocommit=False, autoflush=False, binds={
    # Catalogue
    models.Shop.Catalogue: engines["shop"],
	models.Shop.Sections: engines["shop"],
    models.Shop.Orders: engines["shop"],
    # ...
    # Parts
    models.Parts.Part: engines["parts"],
    models.Parts.BinLocations: engines["parts"],

    # ...
    #Purchasing
    models.Purchasing.SupplierOrder: engines["purchasing"],
    models.Purchasing.SupplierOrder: engines["purchasing"],
    # Company Data
    models.Company.Staffmember: engines["company"],
    models.Company.Suppliers: engines["company"],
    # API Auth
    models.Auth.User: engines["auth"],
    models.Auth.Privileges: engines["auth"],
})

# Dependency
def getDb():
    db = DBSession()
    try:
        yield db
    finally:
        db.close()
```

It's a little laborious having to do this for every model but it does work.

As I have several dbs I thought it would be logical to create a `models` module with sub-files for each db e.g. `models.Parts`, `models.Shop`, `models.Purchase`, `models.Company`, `models.Auth` etc.

/models/__init__.py

``` lang-python

from importlib.metadata import metadata
from sqlalchemy.orm import declarative_base

base = declarative_base()

from . import Auth, Parts, Shop, Catalogue, Purchasing, Shop
```

I can create relationships successfully by importing the `Base` object in the `__init__.py` of `models` and importing that to each sub-file. For example:

### /models/Auth.py ###
``` lang-python
from . import base as Base

from sqlalchemy.orm import relationship
from sqlalchemy import Column, Integer, String, Numeric, Date, DateTime, ForeignKey, null, or_, and_

class User(Base):
    __tablename__ = 'users'

    id = Column(Integer, nullable=False, primary_key=True)
    username = Column(String(256), nullable=False)
    passhash = Column(String(512), nullable=False)
    email = Column(String, nullable=False)
    enabled = Column(Integer, nullable=True)
    staffmember_id = Column(Integer, nullable=False)

    staffmember = relationship("Company.Staffmember", uselist=False)
```

### /models/Company.py ###
``` lang-python
from . import base as Base

from sqlalchemy.orm import relationship
from sqlalchemy import Column, Integer, String, Numeric, Date, DateTime, ForeignKey, null, or_, and_

class Staffmebmer(Base):
    __tablename__ = 'staffmembers'

    id = Column(Integer, ForeignKey("users.staffmember_id"), nullable=False, primary_key=True)
    order = Column(Integer, default=0, nullable=False)
    name = Column(String, nullable=True)
    initial = Column(String, nullable=True)
    email = Column(String, nullable=False)
    enabled = Column(Integer, default=0, nullable=False)

    relationship("Auth.User", back_populates="staffmember")
```

The following route works just fine:

### demo.py ###

``` lang-python

from fastapi import Depends

from sqlalchemy.orm import Session

from .. import app, databases, models

@app.get("/api/user/{id}")
async def read_items(id: int, db: Session=Depends(databases.getDb)):
    user = db.query(models.Auth.User).filter(
        models.Auth.User.id == id
    ).first()

    user.staffmember

    return user
```
Accessing this URL returns:
**(Yes, I'm aware this isn't secure, it is for illustrative purposes only to show that the relationship functions!)**

``` lang-json
{
  "username": "mark",
  "passhash": "<my hash>",
  "enabled": 1,
  "email": "mark@demo.com",
  "id": 1,
  "staffmember_id": 5,
  "staffmember": {
    "order": 20,
    "name": "Mark",
    "email": "mark@demo.com",
    "kStaffmember": 5,
    "initial": "MB",
    "enabled": 1
  }
}
```

However, I want to use steffmember initials as a possible username, so when I qyuery for a user in my OAUTH Authorize scripts I tried to use:

``` lang-python

from ..models import Auth, Company

# 'username' is provided by the auth script from the standard username/password OAuth fields

def get_user(db: Session, username: str):
    db_user_data = db.query(Auth.User).join(Company.Staffmember).filter(
        or_(
            Auth.User.username == username,
            Auth.User.email == username,
            Company.Staffmember.initial == username
        )
    ).first()
```

and I get an Exception:

```
(pymysql.err.ProgrammingError) (1146, "Table 'auth.staffmembers' doesn't exist")
```

Am I going about this whole thing the right way and is there a possible way around this issue?

**Comment by zzzeek** [comment, marked as answer]
you're looking to create a SQL JOIN between two tables "company.staffmembers" and "auth.users", where "company" and "auth" are "legacy databases".

So this problem needs to be approached from a DB / SQL perspective before dealing with any Python code.

When we have "multiple databases" that actually can mean different things particularly with MySQL/MariaDB.   Namely, are these "databases" on different **servers**, or are they two MySQL databases in the same server.

If the two databases were on different servers completely, there's no means of creating a SQL join between those tables unless you made use of a remote database link extension, which AFAICT is not really a thing for MySQL (which is fine, remote database links as we see in Oracle, Postgresql are not easy to work with in any case).

if the two databases are on the **same** server, as seems to be the case here, MySQL/MariaDB allow you to access tables from different databases at once by prepending the table name with the database name, like:

```
select * from company.staffmembers join auth.users on company.staffmembers.auth_id==auth.users.id
```

this also assumes that there is a database account that can connect to MySQL and has GRANT access for those two databases at the same time.  Again, as your program is using the same DB_URL for every database, it looks like that is the case.

but when we write that SQL, it is invoked on a single database connection at a time.   There's not a need to have separate connections that connect to "company", "auth", etc.   We instead just need the table names to render out with their database name prepended.

Where this is going is:

1. lose all those create_engines, make just one engine that connects to any arbitrary database on that MySQL server
2. all the models should have their database name added as their "schema", under `__table_args__`.  See https://docs.sqlalchemy.org/en/14/core/metadata.html?highlight=schema%20name#specifying-the-schema-name and https://docs.sqlalchemy.org/en/14/orm/declarative_tables.html#orm-declarative-table-schema-name
3. SQL will then be rendered for table names including their database names on a particular connection

**Reply by MarkBFamFour** [reply]
Many thanks. I figured there was a more rational way of looking at the database connection than having to bind each individual model to an engine connection. Once I read through your explanation and understood exactly how the approach worked in terms of the SQL equivalent and being on the same database server it made sense.

The only thing that is clear as mud is the documentation on ```schema``` and ```__table_args__``` but I winged it and it seems to work by just adding ```__table_args__ = { "schema": "<database name>" }``` so I guess that's what you were aiming at.

I did find I have to declare every foreign key (even inside the same db/module) explicitly as ```<db>.<table>.<field>``` though, but that makes the foreign key location clearer if nothing else.
